pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -157,9 +160,148 @@ impl Timeline {
                 if !self.cancel.is_cancelled() {
                     tracing::error!("could not compact, repartitioning keyspace failed: {err:?}");
                 }
+                1
             }
         };
 
+        if self.shard_identity.count >= ShardCount::new(2) {
+            // Limit the number of layer rewrites to the number of partitions: this means its
+            // runtime should be comparable to a full round of image layer creations, rather than
+            // being potentially much longer.
+            let rewrite_max = partition_count;
+
+            self.compact_shard_ancestors(rewrite_max, ctx).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Check for layers that are elegible to be rewritten:
+    /// - Shard splitting: After a shard split, ancestor layers beyond pitr_interval, so that
+    ///   we don't indefinitely retain keys in this shard that aren't needed.
+    /// - For future use: layers beyond pitr_interval that are in formats we would
+    ///   rather not maintain compatibility with indefinitely.
+    ///
+    /// Note: this phase may read and write many gigabytes of data: use rewrite_max to bound
+    /// how much work it will try to do in each compaction pass.
+    async fn compact_shard_ancestors(
+        self: &Arc<Self>,
+        rewrite_max: usize,
+        _ctx: &RequestContext,
+    ) -> anyhow::Result<()> {
+        let mut drop_layers = Vec::new();
+        let layers_to_rewrite: Vec<Layer> = Vec::new();
+
+        let layers = self.layers.read().await;
+        for layer_desc in layers.layer_map().iter_historic_layers() {
+            let layer = layers.get_from_desc(&layer_desc);
+            if layer.metadata().shard.shard_count == self.shard_identity.count {
+                // This layer does not belong to a historic ancestor, no need to re-image it.
+                continue;
+            }
+
+            // This layer was created on an ancestor shard: check if it contains any data for this shard.
+            let sharded_range = ShardedRange::new(layer_desc.get_key_range(), &self.shard_identity);
+            let layer_local_page_count = sharded_range.page_count();
+            let layer_raw_page_count = ShardedRange::raw_size(&layer_desc.get_key_range());
+            if layer_local_page_count == 0 {
+                // This ancestral layer only covers keys that belong to other shards.
+                // We do very verbose logging here, including the full metadata: if we had some critical bug that caused
+                // us to incorrectly drop layers, this would simplify manually reinstating those layers.
+                info!(layer=%layer,
+                    "dropping layer after shard split, contains no keys for this shard.  Old metadata: {}",
+                    serde_json::to_string(&IndexLayerMetadata::from(layer.metadata())).unwrap()
+                );
```

Comment:
Why not just debug log it as a parameter?

Reply:
Not sure what you mean: it's intentionally at INFO level because I want it to be recorded when running in prod configs.  I'm using an explicit serde_json call because I want some JSON that I could e.g. scrape out of logs with a script if we had some bad incident.

Reply:
```suggestion
                info!(layer=%layer, ?layer.metadata(),
                    "dropping layer after shard split, contains no keys for this shard.  Old metadata: {}"
                );
```

+-1 the formatting :D

Reply:
On second look, the debug output is basically json equivalent anyway, so fair enough, we can just use `?` formatting: 5d6e18ed5